service/frontend/versionChecker.go

```diff
@@ -0,0 +1,155 @@
+		vc.startOnce.Do(func() {
+			go vc.versionCheckLoop()
+		})
+	}
+}
+
+func (vc *VersionChecker) Stop() {
+	if vc.config.EnableServerVersionCheck() {
+		vc.stopOnce.Do(func() {
+			close(vc.shutdownChan)
+		})
+	}
+}
+
+func (vc *VersionChecker) versionCheckLoop() {
+	timer := time.NewTicker(VersionCheckInterval)
+	defer timer.Stop()
+	vc.performVersionCheck()
+	for {
+		select {
+		case <-vc.shutdownChan:
+			return
+		case <-timer.C:
+			vc.performVersionCheck()
+		}
+	}
+}
+
+func (vc *VersionChecker) performVersionCheck() {
+	sw := vc.metricsScope.StartTimer(metrics.VersionCheckLatency)
+	defer sw.Stop()
+	req, err := vc.createVersionCheckRequest()
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	resp, err := vc.getVersionInfo(req)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	err = vc.saveVersionInfo(resp)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	vc.metricsScope.IncCounter(metrics.VersionCheckSuccessCount)
+}
+
+func (vc *VersionChecker) createVersionCheckRequest() (*check.VersionCheckRequest, error) {
+	clusterMetadataManager := vc.GetClusterMetadataManager()
+	metadata, err := clusterMetadataManager.GetClusterMetadata()
+	if err != nil {
+		return nil, err
+	}
+	return &check.VersionCheckRequest{
+		Product:   headers.ClientNameServer,
+		Version:   headers.ServerVersion,
+		Arch:      runtime.GOARCH,
+		OS:        runtime.GOOS,
+		DB:        clusterMetadataManager.GetName(),
+		ClusterID: metadata.ClusterId,
+		Timestamp: time.Now().UnixNano(),
+	}, nil
+}
+
+func (vc *VersionChecker) getVersionInfo(req *check.VersionCheckRequest) (*check.VersionCheckResponse, error) {
+	return check.NewCaller().Call(req)
+}
+
+func (vc *VersionChecker) saveVersionInfo(resp *check.VersionCheckResponse) error {
+	clusterMetadataManager := vc.GetClusterMetadataManager()
+	metadata, err := clusterMetadataManager.GetClusterMetadata()
+	if err != nil {
+		return err
+	}
+	metadata.VersionInfo = toVersionInfo(resp)
+	saved, err := clusterMetadataManager.SaveClusterMetadata(&persistence.SaveClusterMetadataRequest{
+		ClusterMetadata: metadata.ClusterMetadata, Version: metadata.Version})
+	if err != nil {
+		return err
+	}
+	if !saved {
+		return serviceerror.NewInternal("version info update hasn't been applied")
+	}
+	return nil
+}
+
+func toVersionInfo(resp *check.VersionCheckResponse) *persistenceblobs.VersionInfo {
+	return &persistenceblobs.VersionInfo{
+		Current:      convertReleaseInfo(resp.Current),
+		Recommended:  convertReleaseInfo(resp.Recommended),
+		Instructions: resp.Instructions,
+		Alerts:       convertAlerts(resp.Alerts),
+	}
+}
+
+func convertAlerts(alerts []check.Alert) []*persistenceblobs.Alert {
+	var result []*persistenceblobs.Alert
+	for _, alert := range alerts {
+		result = append(result, &persistenceblobs.Alert{
+			Message:  alert.Message,
+			Severity: int32(alert.Severity),
```

Comment:
Change `alert.Severity` to `int32` in first place.

Reply:
Ok